gobblin-metrics/src/main/java/gobblin/metrics/RootMetricContext.java

```diff
@@ -0,0 +1,142 @@
+import gobblin.metrics.notification.MetricContextCleanupNotification;
+import gobblin.metrics.notification.NewMetricContextNotification;
+import gobblin.metrics.reporter.ContextAwareReporter;
+import gobblin.util.ExecutorsUtils;
+
+
+/**
+ * Special singleton {@link MetricContext} used as the root of the {@link MetricContext} tree. This is the only
+ * {@link MetricContext} that is allowed to not have a parent. Any {@link MetricContext} that does not explicitly
+ * have a parent will automatically become a child of the {@link RootMetricContext}.
+ */
+@Slf4j
+public class RootMetricContext extends MetricContext {
+
+  public static final String ROOT_METRIC_CONTEXT = "RootMetricContext";
+
+  @Getter
+  private final ReferenceQueue<MetricContext> referenceQueue;
+  private final Set<InnerMetricContext> innerMetricContexts;
+  private final ScheduledExecutorService referenceQueueExecutorService;
+
+  private final Set<ContextAwareReporter> reporters;
+
+  private boolean reportingStarted;
+
+  private RootMetricContext(List<Tag<?>> tags) throws NameConflictException {
+    super(ROOT_METRIC_CONTEXT, null, tags, true);
+    this.innerMetricContexts = Sets.newConcurrentHashSet();
+    this.referenceQueue = new ReferenceQueue<MetricContext>();
+    this.referenceQueueExecutorService = MoreExecutors.getExitingScheduledExecutorService(new ScheduledThreadPoolExecutor(1,
+        ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("GobblinMetrics-ReferenceQueue"))));
+    this.referenceQueueExecutorService.scheduleWithFixedDelay(new CheckReferenceQueue(), 0, 2, TimeUnit.SECONDS);
+
+    this.reporters = Sets.newConcurrentHashSet();
+    this.reportingStarted = false;
+  }
+
+  private static void initialize() {
+    try {
+      INSTANCE = new RootMetricContext(Lists.<Tag<?>>newArrayList());
+    } catch (NameConflictException nce) {
+      // Should never happen, as there is no parent, so no conflict.
+      throw new RuntimeException("Failed to generate root metric context. This is an error in the code.");
+    }
+  }
+
+  /**
+   * Get the singleton {@link RootMetricContext}.
+   * @return singleton instance of {@link RootMetricContext}.
+   */
+  public synchronized static RootMetricContext get() {
+
+    if(INSTANCE == null) {
+      initialize();
+    }
+    return INSTANCE;
+  }
+
+  private static RootMetricContext INSTANCE;
+
+  /**
+   * Checks the {@link ReferenceQueue} to find any {@link MetricContext}s that have been garbage collected, and
+   * sends a {@link MetricContextCleanupNotification} to all targets.
+   */
+  private class CheckReferenceQueue implements Runnable {
+    @Override public void run() {
+      Reference<? extends MetricContext> reference;
+      while((reference = referenceQueue.poll()) != null) {
+        ContextWeakReference contextReference = (ContextWeakReference)reference;
+
+        sendNotification(new MetricContextCleanupNotification(contextReference.getInnerContext()));
```

Comment:
I suggest that we catch all RuntimeExceptions and proceed with removal of the context. The catching of RuntimeExceptions is event better to happen in sendNotification() to provide better isolation.